### PR TITLE
[Feature] Add iou3d op from mmdet3d

### PR DESCRIPTION
**Pull request by DCNSW**
## Motivation

Add iou3d cuda operation from mmdet3d (branch: v1.0.0.dev0).

## Modification

Several files in mmcv/ops folder.

## BC-breaking (Optional)

No.

## Use cases (Optional)

from mmcv.ops import boxes_iou_bev, nms_bev, nms_normal_bev

### CODE REVIEW

File: mmcv/ops/iou3d.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    """NMS function GPU implementation (for BEV boxes). The overlap of two
+    boxes for IoU calculation is defined as the exact overlapping area of the
+    two boxes. In this function, one can also set `pre_max_size` and
+    `post_max_size`.
+
+    Args:
+        boxes (torch.Tensor): Input boxes with the shape of [N, 5]
+            ([x1, y1, x2, y2, ry]).
+        scores (torch.Tensor): Scores of boxes with the shape of [N].
+        thresh (int): Threshold.
```
Comment (zhouzaida) [inline mmcv/ops/iou3d.py:40]: `Threshold` is too short. Please explain what `thresh` does.

File: mmcv/ops/iou3d.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    keep = torch.zeros(boxes.size(0), dtype=torch.long)
+    num_out = ext_module.iou3d_nms_forward(boxes, keep, thresh,
+                                           boxes.device.index)
+    keep = order[keep[:num_out].cuda(boxes.device)].contiguous()
+    if post_max_size is not None:
+        keep = keep[:post_max_size]
+    return keep
+
+
+def nms_normal_gpu(boxes, scores, thresh):
```
Comment (zhouzaida) [inline mmcv/ops/iou3d.py:64]: add a unittest for nms_normal_gpu

File: mmcv/ops/csrc/pytorch/cuda/iou3d_cuda.cu [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+                                                   const float *boxes_a,
+                                                   const int num_b,
+                                                   const float *boxes_b,
+                                                   float *ans_overlap) {
+  dim3 blocks(
+      DIVUP(num_b, THREADS_PER_BLOCK),
+      DIVUP(num_a, THREADS_PER_BLOCK));  // blockIdx.x(col), blockIdx.y(row)
+  dim3 threads(THREADS_PER_BLOCK, THREADS_PER_BLOCK);
+
+  iou3d_boxes_overlap_bev_forward_cuda_kernel<<<blocks, threads>>>(
```
Comment (grimoire) [inline mmcv/ops/csrc/pytorch/cuda/iou3d_cuda.cu:25]: Add cuda stream.

File: mmcv/ops/csrc/common/cuda/iou3d_cuda_kernel.cuh [hunk 4]
```diff
@@ -0,0 +1,401 @@
...
+  float left = fmaxf(a[0], b[0]), right = fminf(a[2], b[2]);
+  float top = fmaxf(a[1], b[1]), bottom = fminf(a[3], b[3]);
+  float width = fmaxf(right - left, 0.f), height = fmaxf(bottom - top, 0.f);
+  float interS = width * height;
+  float Sa = (a[2] - a[0]) * (a[3] - a[1]);
+  float Sb = (b[2] - b[0]) * (b[3] - b[1]);
+  return interS / fmaxf(Sa + Sb - interS, EPS);
+}
+
+__global__ void nms_normal_forward_cuda_kernel(const int boxes_num,
```
Comment (grimoire) [inline mmcv/ops/csrc/common/cuda/iou3d_cuda_kernel.cuh:348]: So this is basically 2d nms except boxes is (N,5)?
Can we slice the boxes and use 2d nms directly? Slice does not need much computation.

File: mmcv/ops/csrc/pytorch/iou3d.cpp [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+    iou3d_nms_normal_forward_cuda(boxes_data, mask_data, boxes_num,
+                                  nms_overlap_thresh);
+
+    // unsigned long long mask_cpu[boxes_num * col_blocks];
+    // unsigned long long *mask_cpu = new unsigned long long [boxes_num *
+    // col_blocks];
+    std::vector<unsigned long long> mask_cpu(boxes_num * col_blocks);
+
+    //    printf("boxes_num=%d, col_blocks=%d\n", boxes_num, col_blocks);
+    CHECK_ERROR(cudaMemcpy(&mask_cpu[0], mask_data,
```
[the remark this answers is not in the document]
Comment (DCNSW) [reply mmcv/ops/csrc/pytorch/iou3d.cpp:233]: cudaMemcpy -> cudaMemcpyAsync

File: mmcv/ops/iou3d.py [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+                                           boxes_b.contiguous(), ans_iou)
+
+    return ans_iou
+
+
+def nms_bev(boxes, scores, thresh, pre_max_size=None, post_max_size=None):
+    """NMS function GPU implementation (for BEV boxes). The overlap of two
+    boxes for IoU calculation is defined as the exact overlapping area of the
+    two boxes. In this function, one can also set `pre_max_size` and
+    `post_max_size`.
```
Comment (zhouzaida) [inline mmcv/ops/iou3d.py:34]:
```suggestion
    two boxes. In this function, one can also set ``pre_max_size`` and
    ``post_max_size``.
```

File: mmcv/ops/csrc/pytorch/iou3d.cpp [hunk 7]
```diff
@@ -0,0 +1,262 @@
...
+int iou3d_nms_normal_forward(Tensor boxes, Tensor keep,
+                             float nms_overlap_thresh, int device_id) {
+  // params boxes: (N, 5) [x1, y1, x2, y2, ry]
+  // params keep: (N)
+
+  if (boxes.device().is_cuda()) {
+#ifdef MMCV_WITH_CUDA
+    CHECK_CUDA_INPUT(boxes);
+    CHECK_CONTIGUOUS(keep);
+    cudaSetDevice(device_id);
```
Comment (grimoire) [inline mmcv/ops/csrc/pytorch/iou3d.cpp:208]: Do not use set device, add a device guard instead.

**Comment by zhouzaida** [comment]
please @grimoire have a look.

**Review by grimoire — APPROVED** [review]
LGTM

**Pull request merged**